src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -0,0 +1,259 @@
+                }
+
+                // TODO: Handle requests with body
+
+                state.CancellationToken.ThrowIfCancellationRequested();
+                responseMessage = CreateResponseMessage(requestHandle, state.RequestMessage);
+                state.TrySetResult(responseMessage);
+            }
+            catch (Exception ex)
+            {
+                HandleAsyncException(state, ex);
+            }
+            finally
+            {
+                SafeCurlHandle.DisposeAndClearHandle(ref requestHandle);
+                cancellationTokenRegistration.Dispose();
+            }
+        }
+
+        private void SetOperationStarted()
+        {
+            if (!_anyOperationStarted)
+            {
+                _anyOperationStarted = true;
+            }
+        }
+
+        private SafeCurlHandle CreateRequestHandle(RequestCompletionSource state)
+        {
+            // TODO: If this impacts perf, optimize using a handle pool
+            SafeCurlHandle requestHandle = Interop.libcurl.curl_easy_init();
+            if (requestHandle.IsInvalid)
+            {
+                throw new HttpRequestException(SR.net_http_client_execution_error);
+            }
+
+            Interop.libcurl.curl_easy_setopt(requestHandle, CURLoption.CURLOPT_URL,
+                state.RequestMessage.RequestUri.OriginalString);
+            if (_automaticRedirection)
+            {
+                Interop.libcurl.curl_easy_setopt(requestHandle, CURLoption.CURLOPT_FOLLOWLOCATION, 1);
+            }
+
+            // TODO: Handle headers, proxy and other options
+
+            return requestHandle;
+        }
+
+        private HttpResponseMessage CreateResponseMessage(SafeCurlHandle requestHandle, HttpRequestMessage request)
+        {
+            var response = new HttpResponseMessage();
+            long httpStatusCode = 0;
+            int result = Interop.libcurl.curl_easy_getinfo(requestHandle, CURLINFO.CURLINFO_RESPONSE_CODE,
+                ref httpStatusCode);
+            if (CURLcode.CURLE_OK != result)
+            {
+                throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+            }
+            response.StatusCode = (HttpStatusCode) httpStatusCode;
+
+            // TODO: Do error processing if needed and return actual response
+            response.Content =
+                new StringContent("SendAsync to " + request.RequestUri.OriginalString + " returned: " +
+                                  response.StatusCode);
+            return response;
+        }
+
+        private void HandleAsyncException(RequestCompletionSource state, Exception ex)
+        {
+            if (ex is OperationCanceledException)
+            {
+                // If the exception was due to the cancellation token being canceled, throw cancellation exception.
+                Debug.Assert(state.CancellationToken.IsCancellationRequested);
+                state.TrySetCanceled();
```

Comment:
This should ideally be:
state.TrySetCanceled(((OperationCanceledException)ex).Token);
though with the cast only done once via an as rather than an is.